### ISSUE TITLE
Operations on halfs of a Vector256

### ISSUE THREAD

**Issue by israellot** [dotnet/runtime#76066]
AVX2 operations on x86 are basically two 128 vectors side by side. Currently there's no way to leverage that knowledge when writing Vector256 code.

The bellow example loads 32 bytes and computes the sums of bytes [0..15] and [16..35]. The first one is a handwritten AVX2, the second one is my attempt to replicate that using Vector256. I've included the ASM generated for each. 
At the Vector256 ASM, Vector128 are extracted and summed individually. But the optimal code sums horizontally the Vector256 and extracts the results from upper and lower halfs of it.

```
var zero = Vector256<byte>.Zero;
zero = Avx2.Xor(zero, zero);

var byteVector = Vector256.LoadUnsafe<byte>(ref spanRef);

var shuffled1 = Avx2.UnpackLow(byteVector, zero);
var shuffled2 = Avx2.UnpackHigh(byteVector, zero);

var added = Avx2.Add(shuffled1.AsInt16(), shuffled2.AsInt16());

added = Avx2.HorizontalAdd(added, zero.AsInt16());
added = Avx2.HorizontalAdd(added, zero.AsInt16());
added = Avx2.HorizontalAdd(added, zero.AsInt16());

hash2 = added.GetUpper().AsInt32().ToScalar();
hash1 = added.GetLower().AsInt32().ToScalar();

//ASM

mov      rax, bword ptr [rcx]
vxorps   ymm0, ymm0, ymm0
vmovdqu  ymm1, ymmword ptr[rax]
vpunpcklbw ymm2, ymm1, ymm0
vpunpckhbw ymm1, ymm1, ymm0
vpaddw   ymm1, ymm2, ymm1
vphaddw  ymm1, ymm1, ymm0
vphaddw  ymm1, ymm1, ymm0
vphaddw  ymm1, ymm1, ymm0
vextracti128 xmm0, ymm1, 1
vmovd    eax, xmm0
mov      dword ptr [r8], eax
vmovd    eax, xmm1
mov      dword ptr [rdx], eax

```

```
var byteVector = Vector256.LoadUnsafe<byte>(ref spanRef);
var (low, high) = Vector256.Widen(byteVector);
var added = low + high;
var lower = Vector128.Sum(added.GetLower());
var upper= Vector128.Sum(added.GetUpper());
            
hash1 = lower;
hash2 = upper;

//ASM
mov      rax, bword ptr [rcx]
vmovdqu  ymm0, ymmword ptr[rax]
vmovaps  ymm1, ymm0
vpmovzxbw ymm1, ymm1
vextractf128 xmm0, ymm0, 1
vpmovzxbw ymm0, ymm0
vpaddw   ymm0, ymm1, ymm0
vmovaps  ymm1, ymm0
vphaddw  xmm1, xmm1, xmm1
vphaddw  xmm1, xmm1, xmm1
vphaddw  xmm1, xmm1, xmm1
vmovd    eax, xmm1
movzx    rax, ax
vextracti128 xmm0, ymm0, 1
vphaddw  xmm0, xmm0, xmm0
vphaddw  xmm0, xmm0, xmm0
vphaddw  xmm0, xmm0, xmm0
vmovd    ecx, xmm0
movzx    rcx, cx
mov      dword ptr [rdx], eax
mov      dword ptr [r8], ecx

```

One clear improvement would be the JIT detecting the same operation is being done to the lower and upper half of a vector 256 and emitting an AVX2 256 wide operation and extracting the results afterwards.

```
var lower = Vector128.Sum(v.GetLower());
var upper= Vector128.Sum(v.GetUpper());
```

Second option would be new apis that operate at the halfs.
`(Vector128 low Vector128 high) Vector256.SumHalfs()`

**Comment by Gnbrkm41** [comment]
Note: There exists a numeric type called `System.Half`, naming such a method `SumHalfs` may confuse people...

> AVX2 operations on x86 are basically two 128 vectors side by side. Currently there's no way to leverage that knowledge when writing Vector256 code.

From my knowledge, `Vector` helper methods are introduced to remove bulk of the duplicated code in vectorised codes where many logics that does not require underlying implementation of the internals are duplicated, such as `LoadUnsafe` and `Add` - simple, common operations that operates with pairs of elements from two vectors, not elements across a single vector - previously, to write vectorised codes that works across multiple ISAs, you had to do:
```cs
if (Sse2.IsSupported)
{
    Vector128<byte> vector = Sse2.LoadVector128((byte*)ptr + i);
    /* code using ISA-specific intrinsics... */
}
else if (AdvSimd.IsSupported)
{
    Vector128<byte> vector = AdvSimd.LoadVector128((byte*)ptr + i);
    /* code using ISA-specific intrinsics... */
}
```
whereas now it just boils down to:
```cs
if (Vector128.IsSupported)
{
    Vector128<byte> vector = Vector128.LoadUnsafe(ref ...)
}
```
...making code much more cleaner by abstracting away some of the details so you can share most of the shared logics, however you are also supposed to use ISA-specific intrinsics where necessary in case there is a more specific instruction available in a specific ISA or if disassembly/profiling of such code results in a suboptimal code generation due to the abstraction provided.

In other words, Vector helper methods are made to replace many of the common duplicate operations but is *not meant to replace the entirety of ISA-specific instructions*. It is purely x64's implementation of 256-bit SIMD instructions that treats 256-bit vectors as two 128-bit vectors logically, and even in x64, not all vector operations consider the vector to be two lanes of separate 128-bit vectors, as there's instructions that operate across the entire 256-bit vector. It doesn't make sense to add a method reflecting ISA-specific implementation back to a generalized helper class for ISA-agnostic vector operations.

**Comment by Gnbrkm41** [comment]
> The below example loads 32 bytes and computes the sums of bytes [0..15] and [16..31]

The example here already takes advantage of the ISA-specific detail of AVX2 that internally many instructions operate on it as a pair of 128-bit vectors, using it to calculate sums of individual 128-bit sections; A more generalised approach would be to take a single 128-bit vector at a time and calculating the sum.

**Comment by JulieLeeMSFT** [comment]
Cc @tannergooding @BruceForstall.

**Comment by tannergooding** [comment]
@Gnbrkm41 has it correct.

If you want to operate on 2x 128-bit vectors, I'd recommend explicitly dealing with the underlying V128 types. It's possible the JIT could minimally recognize this and change it to do `1x 256-bit op` internally, but I expect the benefits of this would be exceedingly rare, especially when considering what is "best" for cross-platform code.

**Comment by BruceForstall** [comment]
Based on Tanner's answer, I'm going to go ahead and close this.
